rfcs/003-workspaces.md

```diff
@@ -0,0 +1,279 @@
+            "starter_code.h",
+            "starter_code.c"
+        ]
+    }
+}
+```
+
+The home directory in the workspace will be determined by the `workspace` directory inside a question directory. In the future, we'll add the ability to dynamically generate files via `server.py` and place them into the home directory. This is not part of the MVP.
+
+```
+questions
+`-- myquestion
+    +-- info.json
+    +-- question.html
+    +-- server.py
+    +-- clientFilesQuestion
+    +-- tests
+    |   +-- correct_answer.c
+    |   `-- test_run.py
+    |
+    `-- workspace
+        +-- .bashrc
+        +-- starter_code.h
+        `-- starter_code.c
+```
+
+The workspace options will need to be synced to the `questions` table via the usual syncing code.
+
+### Student-facing question interface
+
+> What happens when we render a question with an associated workspace?
+
+When a new variant of a question is created, the main server will create a corresponding workspace in the database associated with that particular variant. This database entry will contain a unique hash/id/something. However, we're not going to actually provision any containers, etc. for this workspace just yet. The `state` of the new workspace row will be `uninitialized`.
+
+We'll introduce a new `<pl-workspace>` element that renders (to start with) a "Launch workspace" button. We should introduce a new `workspace_url` to `data.options`, and this element (or potentially other elements) can use this to render a button. `workspace_url` will be something of the form `/workspace/[workspace_id]`.
+
+> Implementation notes:
+> * Workspace will be created and inserted into the DB in the `variants_insert` sproc
+> * In `variants_insert`, query for `questions.workspace_image` to determine if it's necessary to create a workspace
+> * `workspaceUrl` should be generated in the `_buildQuestionUrls` function in `lib/question.js`
+> * `data.options.workspace_url` should be populated in the `renderPanel` function of `question-servers/freeform.js`; `workspaceUrl` should be available from the `locals` object like the other URLs
+> * `<pl-workspace>` element should read `data.options.workspace_url` and render a block-level link styled as a button that opens the workspace URL in a new tab
+
+When this button is clicked, the URL at `workspace_url` will be opened in a new tab.
+
+### Accessing a workspace
+
+> What happens when a user lands on a `workspace_url`?
+
+`workspace_url` pages will be served by the main PrairieLearn server (someday, we could split this into a separate autoscaled component).
+
+When the main server gets a request to this url, we'll first check if we have an existing instance of a workspace by checking the `state` column of the appropriate workspace. There will be three cases here:
+
+#### Workspace is in state `uninitialized`
+
+* Respond to the request with the basic markup for the outer frame
+* Begin a transaction
+  * Obtain a lock on the row in the `workspaces` table
```

Comment:
How do we lock the row? I assume it involves one of these, but not sure exactly:
* wrap the query in a `sqldb.beginTransaction`
* wrap with `namedLocks`
* create+call a `workspace_lock` sproc